### ISSUE TITLE
Limitations in content scripts

### ISSUE THREAD

**Issue by Nickwiz** [mdn/content#25821]
### MDN URL

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/executeScript

### What specific section or headline is this issue about?

Limitations

### What information was incorrect, unhelpful, or incomplete?

Believe it would be helpful to include a *limitations* section listing various limitations for content-scripts.

### What did you expect to see?

Was porting a normal script bundle to extension but got stuck at several turns.

1) `import` i.e. use of modules is not supported. E.g. one can not split up content scripts in modules. Suspect all due to how the scripts are injected into one environment. Guess one might could do a work-around by creating script nodes that one apply to the DOM but that gain likely defeat isolation.

2) `customElements` is not accessible. I.e. one can not define new custom elements.

Sure there are more, but from the onset it seems logical at least `customElements` should be accessible. Instead one end up with cryptic errors. Some sort of short list of limitations would be helpful so that one do not have to test, fail, debug, look around the web, ... when a bullet-list could have prevented this moot route.

### Do you have any supporting links, references, or citations?

https://stackoverflow.com/questions/42800035/why-cant-you-create-custom-elements-in-content-scripts

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/mozilla/add-ons/webextensions/api/tabs/executescript`
* MDN URL: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/executeScript
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/mozilla/add-ons/webextensions/api/tabs/executescript/index.md
* Last commit: https://github.com/mdn/content/commit/3efc65cf4477776f8da581c5f2d99e58f53b3649
* Document last modified: 2023-03-22T03:16:23.000Z

</details>

**Comment by guest271314** [comment]
Can you elaborate on `import` usage? Did you try dynamic `import()`?

**Comment by Nickwiz** [comment]
> Can you elaborate on `import` usage? Did you try dynamic `import()`?

No, I used the [`declaration-style`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import)
Using dynamic import `import()` works as advertised.

**Comment by guest271314** [comment]
Note, technically `scripting.executeScript()` is not a content script.

**Comment by Rob--W** [comment]
Static imports are not supported yet:

- https://bugzil.la/1451545 "Bug 1451545 - Support loading content scripts as ES6 modules"
- https://crbug.com/783757 "Issue 783757: module script support for content scripts and chrome.tabs.executeScript"

Dynamic imports work, though with some caveats: The script must be in web_accessible_resources (if in MV3) and absolute path should be passed. Relevant bugs:
- https://bugzil.la/1803950 "Bug 1803950 - Dynamic import fails in content script in MV3 extension"
- https://crbug.com/1247760 "Issue 1247760: Chrome Extension content script dynamic import origin change"

Custom elements as a feature are not well-supported in content scripts. The reasons vary, but there is a fundamental issue in the design of Custom Elements and the isolation of content scripts. Usually, web pages cannot directly interact with JS objects defined in a content script. A custom element registration consists of registering a class & constructor for the custom element, and any custom element instance is a subclass of the given element. The usual expectation is an instantiated custom element to be an instance of that class, but given the isolation of the JS "worlds", that would be infeasible. The issue can be overcome by registering the custom element in the context of the web page. Here are some of the relevant bugs:

- https://bugzil.la/1771291 "Bug 1771291 - Can't define new custom element after a previous one was created, in Add-On Content Scripts"
- https://crbug.com/390807 "Issue 390807: Content scripts can't define custom elements"

**Comment by bacharakis** [comment]
I am closing this issue, given that Rob provided information and feedback. @Nickwiz, if you think your initial request isn't covered and we should update the docs, I'd be happy to reopen it with updated suggestions. Thank you!
